File: app/src/jvx/java/art/coded/wireframe/viewmodel/HomeViewModel.java
package art.coded.wireframe.viewmodel;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.preference.PreferenceManager;

import art.coded.wireframe.R;

public class HomeViewModel extends ViewModel {

    private static final String LOG_TAG = HomeViewModel.class.getSimpleName();

    private final MutableLiveData<String> mText;

    public HomeViewModel() {
        mText = new MutableLiveData<>();
    }

    public void loadData(Context context) {

        String username = PreferenceManager.getDefaultSharedPreferences(context)
                .getString(context.getString(R.string.key_username),
                        context.getString(R.string.default_username));

        mText.setValue(String.format("Hello, %s!", username));
    }

    public LiveData<String> getText() {
        return mText;
    }
}

File: app/src/ktx/java/art/coded/wireframe/viewmodel/HomeViewModel.kt
package art.coded.wireframe.viewmodel

import androidx.lifecycle.LiveData

import android.content.SharedPreferences
import androidx.lifecycle.ViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider

private val LOG_TAG = HomeViewModel::class.java.simpleName

class HomeViewModel(private val sharedPreferences: SharedPreferences,
                    private val key: String,
                    private val defaultVal: String)
    : ViewModel() {

    val actionPending = MutableLiveData<Boolean>()

    fun getData(): LiveData<String> {
        actionPending.postValue(true)
        val username = sharedPreferences.getString(key, defaultVal)
        val text = MutableLiveData<String>()
        text.value = String.format("Hello, %s!", username)
        actionPending.postValue(false)
        return text
    }

    class HomeViewModelFactory(private val sharedPreferences: SharedPreferences,
                               private val key: String,
                               private val defaultVal: String) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return HomeViewModel(sharedPreferences, key, defaultVal) as T
        }
    }
}